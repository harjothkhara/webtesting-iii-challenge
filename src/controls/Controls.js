import React from 'react';

const Controls = props => {
  const { locked, closed, toggleLocked, toggleClosed } = props;

  return (
    <div className="controls panel">
      {/* button is clicked (toggleLocked function invoked, closed=true and state changes: locked=true) then display 'unlock gate' if locked is false, otherwise, display lock gate.  */}
      <button disabled={!closed} onClick={toggleLocked} className="toggle-btn">
        {/* if button is clicked, gate -> locked (true), text dynamically changes to display unlock gate, otherwise if not clicked, default state conditions remain in place of locked=false, closed=true, in this case display locked gate */}
        {locked ? 'Unlock Gate' : 'Lock Gate'}
      </button>

        {/* button is clicked (toggleClosed function invoked, locked=false and state changes: closed=true) then display open gate if closed is false, otherwise, display close gate.  */}
      <button disabled={locked} onClick={toggleClosed} className="toggle-btn">
      {/* if button is clicked, gate -> closed (true), text dynamically changes to display open gate, otherwise if not clicked, default state conditions remain in place of locked=false, closed=false, in this gate display close gate. */}
        {closed ? 'Open Gate' : 'Close Gate'} 
      </button>
    </div>
  );
};

export default Controls;
