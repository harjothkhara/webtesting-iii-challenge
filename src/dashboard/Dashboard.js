import React from 'react';

import Display from '../display/Display';
import Controls from '../controls/Controls';

class Dashboard extends React.Component {
  state = {
    locked: false,
    closed: false,
  };

  render() {
    const { closed, locked } = this.state;

    return (
      <>
        {/* passing default state down as props */}
        <Display locked={locked} closed={closed} />
        {/* passing default state and toggle functions as props */}
        <Controls
          locked={locked}
          closed={closed}
          toggleLocked={this.toggleLocked}
          toggleClosed={this.toggleClosed}
        />
      </>
    );
  }

// Toggle Functions

  toggleLocked = () => {
    this.setState(prev => ({ locked: !prev.locked }));
  // updating locked to true based on the previous state condition
  };

  toggleClosed = () => {
    this.setState(prev => ({ closed: !prev.closed }));
    // updating closed to true based on the previous state condition
  };
}

export default Dashboard;