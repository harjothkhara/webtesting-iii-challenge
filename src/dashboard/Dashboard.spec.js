import React from 'react';
import { render, fireEvent } from "@testing-library/react";
import '@testing-library/react/cleanup-after-each';
import "jest-dom/extend-expect";

import Dashboard from "./Dashboard";
import Controls from '../controls/Controls';

describe("<Dashboard />", () => {
    it("renders successfully", () => {
        render(<Dashboard />);
    });
    it('matches the snapshot', () => {
        const { container }= render(<Dashboard />); 
        expect(container.firstChild).toMatchSnapshot(); 
    });
    //making sure that controls and displays render
    it('renders controls and display', () => {
        render(<Dashboard />);
        // everything that is not 0, “”, null, undefined, NaN or false is truthy.
        expect(document.querySelector('.controls')).toBeTruthy();
        expect(document.querySelector('.display')).toBeTruthy(); 
    });

    it("gate opens and closes successfully when unlocked", () => {
        const { getByText } = render(<Dashboard />);
        const gate = getByText(/Close/); 
        fireEvent.click(gate);
        fireEvent.click(gate);
        fireEvent.click(gate);
        fireEvent.click(gate);
        getByText(/Open/);
    });
    it("gate locks successfully", () => {
        const { getByText } = render(<Dashboard />);
        const lock = getByText(/Lock/);
        const gate = getByText(/Close/);
        fireEvent.click(gate);
        fireEvent.click(lock);
        getByText(/Locked/);
        getByText(/Closed/);
    });
    it("gate locks and unlocks successfully", () => {
        const { getByText } = render(<Dashboard />);
        const lock = getByText(/Lock/);
        const gate = getByText(/Close/);
        fireEvent.click(gate);
        fireEvent.click(lock);
        fireEvent.click(lock);
        getByText(/Unlocked/);
        getByText(/Closed/);
    });
    it("locks only when the gate is closed, opens only when unlocked", () => {
        const { getByText } = render(<Dashboard />);
        const lock = getByText(/Lock/);
        const gate = getByText(/Close/);
        fireEvent.click(lock);
        fireEvent.click(gate);
        fireEvent.click(lock);
        fireEvent.click(gate);
        getByText(/Locked/);
        getByText(/Closed/);
    })
    
});