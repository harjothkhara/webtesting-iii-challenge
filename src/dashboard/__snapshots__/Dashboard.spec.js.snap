// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Dashboard /> matches the snapshot 1`] = `
<div
  class="display panel"
>
  <div
    class="led green-led"
  >
    Unlocked
  </div>
  <div
    class="led green-led"
  >
    Open
  </div>
</div>
`;
